runtime/weak.c

```diff
@@ -358,20 +358,28 @@ CAMLprim value caml_ephe_get_data (value ar)
   return optionalize(caml_ephemeron_get_data(ar, &data), &data);
 }
 
-
-Caml_inline void copy_value(value src, value dst)
-{
-  if (Tag_val (src) < No_scan_tag){
-    mlsize_t i;
-    for (i = 0; i < Wosize_val (src); i++){
-      value f = Field (src, i);
-      if (caml_gc_phase == Phase_mark && Must_be_Marked_during_mark(f)){
-        caml_darken (f, NULL);
-      }
-      caml_modify (&Field (dst, i), f);
+static void copy_value(value src, value dst)
+{
+  mlsize_t sz, i;
+  sz = Wosize_val(src);
+  if (Tag_val (src) >= No_scan_tag) {
+    /* Direct copy */
+    memcpy (Bp_val (dst), Bp_val (src), Bsize_wsize (sz));
```

Comment:
Is it intentional to turn the `memmove` into a `memcpy` here? It is not immediately obvious looking at use-sites of `copy_value` that we never have `dst == src`.

Reply:
It is intentional because I expect `dst` to be freshly-allocated.

To me, `memmove` is a code smell.  With a Rust-style coding discipline you should never copy something into itself.

Reply:
> It is intentional because I expect `dst` to be freshly-allocated.

This sounds reasonable given the `weak.c` functions, where the `elt` value used as destination is either `Val_unit` (in which case the function is not called) or the result of a `caml_alloc` call. (Note: those only callers also guarantee that the tag is not Custom_tag or Infix_tag; this precondition could be added as an assert.)